osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs

```diff
@@ -19,46 +20,155 @@ namespace osu.Game.Rulesets.Taiko.Difficulty
 {
     public class TaikoDifficultyCalculator : DifficultyCalculator
     {
-        private const double star_scaling_factor = 0.04125;
+        private const double rhythm_skill_multiplier = 0.014;
+        private const double colour_skill_multiplier = 0.01;
+        private const double stamina_skill_multiplier = 0.02;
 
         public TaikoDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
             : base(ruleset, beatmap)
         {
         }
 
+        protected override Skill[] CreateSkills(IBeatmap beatmap) => new Skill[]
+        {
+            new Colour(),
+            new Rhythm(),
+            new Stamina(true),
+            new Stamina(false),
+        };
+
+        protected override Mod[] DifficultyAdjustmentMods => new Mod[]
+        {
+            new TaikoModDoubleTime(),
+            new TaikoModHalfTime(),
+            new TaikoModEasy(),
+            new TaikoModHardRock(),
+        };
+
+        protected override IEnumerable<DifficultyHitObject> CreateDifficultyHitObjects(IBeatmap beatmap, double clockRate)
+        {
+            List<TaikoDifficultyHitObject> taikoDifficultyHitObjects = new List<TaikoDifficultyHitObject>();
+
+            for (int i = 2; i < beatmap.HitObjects.Count; i++)
+            {
+                // Check for negative durations
+                var currentAfterLast = beatmap.HitObjects[i].StartTime > beatmap.HitObjects[i - 1].StartTime;
+                var lastAfterSecondLast = beatmap.HitObjects[i - 1].StartTime > beatmap.HitObjects[i - 2].StartTime;
```

Comment:
This should be impossible. Was this just something added for safety, or did you actually observe this happening in any beatmaps, and if so do you know which ones?

If this is possible then the fix should be elsewhere, because it affects all rulesets.

Reply:
I'm almost certain I haven't added this conditional myself, just retouched it slightly. @jgdhs27 should probably be able to respond to this better.

Reply:
So I think this existed originally to fix some NaN values being output, but I've tested it across 93k ranked/loved beatmaps and all looks fine.

I think this was probably fixed via https://github.com/ppy/osu/pull/9573/commits/1116703e92c1a6ac82d4b63298780d5ea584ace8 (this particular change was added in https://github.com/ppy/osu/pull/9294/commits/5852a37eb7499ac3969def1845fd3e2115f3236d, prior to that commit).

Will remove.